suite/tests/client-interface/drstatecmp-fuzz-app.c

```diff
@@ -0,0 +1,238 @@
+#include <stdint.h>
+#include <stdio.h>
+#include <time.h>
+
+#define NUM_INSTS 10000
+
+#define VERBOSE 0
+
+#ifdef AARCH64
+
+static byte *generated_code;
+static size_t code_size;
+
+static jmp_buf mark;
+static sigjmp_buf sig_mark;
+
+void
+sig_segv_fpe_handler(int signal)
+{
+    siglongjmp(sig_mark, 1);
+    DR_ASSERT_MSG(false, "Shouldn't be reached");
+}
+
+void
+sigill_handler(int signal, siginfo_t *siginfo, void *uctx)
+{
+    ucontext_t *context = (ucontext_t *)uctx;
+    /* Skip illegal instructions. */
+    context->uc_mcontext.pc += 4;
+}
+
+static void
+print_instr_pc(instr_t *instr, byte *encode_pc)
+{
+    fprintf(stderr, "%p: ", encode_pc);
+    instr_disassemble(GLOBAL_DCONTEXT, instr, STDERR);
+    fprintf(stderr, "\n");
+}
+
+static byte *
+append_instr(instr_t *instr, byte *encode_pc)
+{
+#    if VERBOSE > 0
+    print_instr_pc(instr, encode_pc);
+#    endif
+    byte *nxt_pc = instr_encode(GLOBAL_DCONTEXT, instr, encode_pc);
+    assert(nxt_pc);
+    instr_destroy(GLOBAL_DCONTEXT, instr);
+    return nxt_pc;
+}
+
+static uint32_t
+rand_32b(void)
+{
+    uint32_t r = rand() & 0xff;
+    r |= (rand() & 0xff) << 8;
+    r |= (rand() & 0xff) << 16;
+    r |= (rand() & 0xff) << 24;
+    return r;
+}
+
+static uint32_t
+generate_encoded_inst(void)
+{
+    /* Pick one of the available (side-effect-free and non-branch) opcodes
+     * and randomize the non-fixed bits.
+     */
+    uint32_t opcode_pick = rand() % FUZZ_INST_CNT;
+    const opcode_opnd_pair_t opcode_opnd_pair = fuzz_opcode_opnd_pairs[opcode_pick];
+
+    uint32_t opnd_mask = opcode_opnd_pair.opnd;
+    /* Avoid registers 16-31 for Rd (special registers and callee-saved registers) to
+     * avoid causing segmentation faults and skipping checks.
+     * XXX: Could relax this constraint by pushing all callee-saved registers in a new
+     * preceding basic block and popping them in a succeeding basic block (need
+     * to be separate basic blocks since drstatecmp does not support memory operations).
+     * Encoding for register 31 for Rd should still be avoided since this encoding is used
+     * for special purposes and sometimes refers to the stack pointer or the zero
+     * register.
+     */
+    opnd_mask &= 0b01111;
+    /* Fuzz the operand bits. */
+    uint32_t fuzzed_opnd = rand_32b() & opnd_mask;
+
+    uint32_t encoded_inst = opcode_opnd_pair.opcode | fuzzed_opnd;
+    return encoded_inst;
+}
+
+static int
+check_decoded_inst(instr_t *decoded_inst)
+{
+    return instr_valid(decoded_inst) && instr_get_opcode(decoded_inst) != OP_xx &&
+        instr_raw_bits_valid(decoded_inst) && instr_operands_valid(decoded_inst);
+}
+
+static byte *
+generate_inst(byte *encode_pc)
+{
+    /* Pick a random side-effect-free and non-branch instruction. */
+    uint32_t encoded_inst = generate_encoded_inst();
+    byte encoded_inst_bytes[4];
+    for (int i = 0; i < 4; ++i, encoded_inst >>= 8)
+        encoded_inst_bytes[i] = (byte)(encoded_inst & 0xff);
```

Comment:
But the encoding is already in `encoded_inst` in little-endian form: this byte array and loop can be removed and we can just point at `&encoded_inst`.

Reply:
Right. Wasn't sure about the endianness and did the byte array to be safe but it is simpler this way. Changed.